crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -4935,22 +5843,357 @@ impl CircuitHandle {
                         elapsed.as_secs_f64()
                     );
                 }
-            });
-        Ok(())
+
+                Ok(())
+            })
     }
 
-    pub fn restore(&mut self, base: &StoragePath) -> Result<(), SchedulerError> {
-        self.circuit
-            .map_nodes_recursive_mut(&mut |node: &mut dyn Node| {
-                node.restore(base).expect("restored");
-            });
+    /// Restores the circuit from a checkpoint.
+    ///
+    /// Restore the circuit from a checkpoint and prepare it to backfill new and
+    /// mofified parts of the circuit if necessary.
+    ///
+    /// 1. Find and restore the checkpointed state of each operator.
+    /// 2. Identify stateful operators (such as integrals and output nodes) that don't have
+    ///    a checkpoint and require backfill (the `need_backfill` set).
+    /// 3. Iterate backward from `need_backfill` nodes to find all operators that
+    ///    should participate in the replay phase of the circuit. Iteration stops when
+    ///    reaching a stream whose contents can be replayed from an existing node
+    ///    that has a checkpoint or an input node.
+    /// 4. If the circuit requires backfill, prepare the circuit for replay by
+    ///    configuring the scheduler to only schedule nodes that participate in
+    ///    backfill.
+    ///
+    /// Returns `None` if the circuit does not require backfill; returns info about
+    /// nodes that participate in backfill otherwise.
+    ///
+    /// * After calling this function, the client can invoke `step` repeatedly for replay to make progress.
+    /// * Use `is_replay_complete` to determine whether the circuit has finished the replay.
+    /// * Use `complete_replay` to finalize the replay phase and prepare the circuit for normal operation.
+    pub fn restore(&mut self, base: &StoragePath) -> Result<Option<BootstrapInfo>, DbspError> {
+        // Nodes that will act as replay sources during the replay phase of the circuit.
+        let mut replay_sources: BTreeMap<NodeId, StreamId> = BTreeMap::new();
+
+        // Nodes that require backfill from upstream nodes.
+        let mut need_backfill: BTreeSet<GlobalNodeId> = BTreeSet::new();
+
+        // debug!("CircuitHandle::restore: restoring from checkpoint {}", base);
```

Comment:
why do you need to comment this out?

Reply:
This code runs in every worker thread, so will generate the output N times. This is why I prefer to do most logging in the `dbsp_handle`, which is one thread talking to all workers.